Tuples: adding more tests for IDE scenarios

jcouv opened the pull request:
I'm going down the list of folders in `CSharpEditorServicesTest` and adding IDE tests for tuples for all the relevant refactorings. I skipped over `Completion` and taking a break at `EditAndContinue`. Additional test suggestions within this range are appreciated.

There are four spots marked as "note" in comments. I plan to remove those before merging, but I want to discuss the expectations on this thread (leave as-is, simple fix, or file issue).

The only non-test code is in two files (`src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs`, `src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs`).

@dotnet/roslyn-ide and @VSadov for review.
I'll get sign-off from compiler team once the rest is sorted out.

Fixes #10566.

jcouv commented on src/EditorFeatures/CSharpTest/CodeActions/GenerateFromMembers/AddConstructorParameters/AddConstructorParametersTests.cs:
```diff
@@ -67,5 +68,58 @@ protected override object CreateCodeRefactoringProvider(Workspace workspace)
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddConstructorParameters), Trait(Traits.Feature, Test.Utilities.Traits.Features.Tuples)]
+        public async Task TestTupleOptionalWithNames()
+        {
+            // Note: default((string c, int d)) is less than ideal
```
// Note: default((string c, int d)) is less than ideal

jcouv commented on src/EditorFeatures/CSharpTest/CodeActions/GenerateFromMembers/GenerateEqualsAndGetHashCode/GenerateEqualsAndGetHashCodeTests.cs:
```diff
@@ -175,5 +176,54 @@ public override bool Equals(object obj)
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateEqualsAndGetHashCode), Trait(Traits.Feature, Test.Utilities.Traits.Features.Tuples)]
+        public async Task TupleWithNames_Equals()
+        {
+            // Note: EqualityComparer<(int x, string y)> is less than ideal
```
// Note: EqualityComparer<(int x, string y)> is less than ideal

jcouv commented on src/EditorFeatures/CSharpTest/CodeActions/GenerateFromMembers/GenerateEqualsAndGetHashCode/GenerateEqualsAndGetHashCodeTests.cs:
```diff
@@ -175,5 +176,54 @@ public override bool Equals(object obj)
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateEqualsAndGetHashCode), Trait(Traits.Feature, Test.Utilities.Traits.Features.Tuples)]
+        public async Task TupleWithNames_HashCode()
+        {
+            // Note: EqualityComparer<(int x, string y)> is less than ideal
```
// Note: EqualityComparer<(int x, string y)> is less than ideal

jcouv commented on src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs:
```diff
@@ -3950,5 +3951,102 @@ static void N()
+{
+    public void M()
+    {
+        (((int a, string b))((c: (int)1, d: (string)""hello""))).a.ToString();
```
This seem like it could be simplified.

DustinCampbell replied:
That'll take work in the simplifier. Feel free to file an issue if you haven't already.

jcouv replied:
Opened #11028 for simplifier work.

balajikris replied:
Also, add a `[WorkItem(11028, "url")]` attribute to this test.

jcouv replied:
Added.

jcouv commented on src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs:
```diff
@@ -2668,5 +2669,158 @@ class C
+                index: 2,
+                parseOptions: TestOptions.Regular.WithTuplesFeature())
+                );
+        }
```
Is there a better way to check that refactorings with index 0 and 1 are available, but not 2 or 3?

balajikris replied:
you can test for the number of fixes/refactorings offered. So in this case, it will be 2 instead of 4 offered. I suppose you can also test the display string of offered fixes and assert introduce constant is not offered.

jcouv commented on src/EditorFeatures/CSharpTest/Diagnostics/GenerateMethod/GenerateMethodTests.cs:
```diff
@@ -2879,84 +2880,31 @@ private static void OnChanged(object sender, EventArgs e)
 }");
         }
 
-        public class GenerateConversionTest : AbstractCSharpDiagnosticProviderBasedUserDiagnosticTest
```
This was a nested class which accidentally caused me grief. I moved it to a separate file.

jcouv commented on src/EditorFeatures/CSharpTest/Diagnostics/GenerateMethod/GenerateConversionTests.cs:
```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
This file is just moving `GenerateConversionTest` to its own file without code changes.

jcouv commented on src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs:
```diff
@@ -2668,5 +2669,158 @@ class C
+{
+    private static readonly (int a, string Item2) {|Rename:p|} = (a: 1, ""hello"");
+    var i = p.ToString() + p.ToString();
+}";
```
Mads suggested the refactoring of tuples with missing element names should generate a nameless tuple, rather than fill in the missing names with the default names (`Item1`, `Item2`, ...). If you agree, I will file an issue.

jcouv commented:
@dotnet/roslyn-ide Please review. Thanks

jcouv commented:
@dotnet/roslyn-ide This PR needs some review love. Thanks

balajikris commented on src/EditorFeatures/CSharpTest/BraceHighlighting/BraceHighlightingTests.cs:
```diff
@@ -179,5 +171,12 @@ void M(int variable)
+        public async Task TestTuples()
+        {
+            await TestBraceHighlightingAsync(@" class C { [|(|]int, int[|)|]$$ x = (1, 2); } ", TestOptions.Regular.WithTuplesFeature());
+            await TestBraceHighlightingAsync(@" class C { (int, int) x = [|(|]1, 2[|)|]$$; } ", TestOptions.Regular.WithTuplesFeature());
```
test for nested braces? `(Foo<T>, int)`, `((int,int), int)`. I haven't exactly read the specification to know if these are legal syntax. If they are not, they can at the least be negative tests :smile:

jcouv replied:
Yes, it is valid. I will add test.

jcouv replied:
Done.

balajikris commented on src/EditorFeatures/CSharpTest/BraceMatching/CSharpBraceMatcherTests.cs:
```diff
@@ -747,5 +749,41 @@ static void Main(string[] args)
+            var expected = @"public class C { var x = [|(|]1, 2, 3, 4, 5, 6, 7, 8); }";
+
+            await TestAsync(code, expected, TestOptions.Regular.WithTuplesFeature());
+        }
```
similarly, nested brace cases, if you see any value in them.

jcouv replied:
Added.

balajikris commented on src/EditorFeatures/CSharpTest/CodeActions/GenerateFromMembers/AddConstructorParameters/AddConstructorParametersTests.cs:
```diff
@@ -67,5 +68,58 @@ protected override object CreateCodeRefactoringProvider(Workspace workspace)
+            await TestMissingAsync(
+@"class Program { [|(int a, string b) i ; (string c, int d) s ;|] public Program ( (int e, string f) i ) { this . i = i ; } } ",
+parseOptions: TestOptions.Regular.WithTuplesFeature());
+        }
```
are nullable types and generic types legal in tuples? if so, they may make interesting tests.

jcouv replied:
Yes, you can do `(int?, int?)` or `(List<int>, List<int>)`. I'll add.

jcouv replied:
Done.

balajikris commented on src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs:
```diff
@@ -2668,5 +2670,158 @@ class C
+@"class C
+{
+    var i = [|(1, ""hello"")|].ToString();
+}";
```
what happens if i select individual items that form this tuple and apply the same refactoring. does it generate valid syntax? or should that be an invalid selection and the refactoring not be offered? 
E.g: Select `1` and invoke Introduce variable.

jcouv replied:
Added.

balajikris commented on src/EditorFeatures/CSharpTest/Diagnostics/GenerateConstructor/GenerateConstructorTests.cs:
```diff
@@ -785,5 +784,83 @@ public bool M(string s, int i, bool b)
+@"class C { void M() { new [|C|]((1, ""hello"")); } }",
+@"class C { private (int, string) p; public C((int, string) p) { this.p = p; } void M() { new C((1, ""hello"")); } }",
+parseOptions: TestOptions.Regular.WithTuplesFeature());
+        }
```
may be test for a constructor that takes other types in addition to a tuple type? `new C((1, true), "hello")`

jcouv replied:
Added a bool parameter to this constructor refactoring.

VSadov commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -567,7 +567,7 @@ public override bool IsReferenceType
             get
             {
-                return _underlyingType.IsReferenceType;
+                return _underlyingType.IsErrorType() ? false : _underlyingType.IsReferenceType;
```
Is there a test for this?

jcouv replied:
Yes. Actually almost all the IDE tests hit this because the code doesn't include code or references for `ValueTuple`. I was wondering whether this should just always return `false` assuming we also add a check that the underlying type is a struct. Is there any reason for the tuple to be a reference type?

jcouv replied:
I'll add a compiler test. Thanks

balajikris commented on src/EditorFeatures/CSharpTest/Diagnostics/UseImplicitOrExplicitType/UseImplicitTypeTests.cs:
```diff
@@ -1441,5 +1442,24 @@ static void M()
+@"class C { static void M() { [|(int, string)|] s = (c: 1, d: ""hello""); } }",
+options: ImplicitTypeEverywhere(),
+parseOptions: TestOptions.Regular.WithTuplesFeature());
+        }
```
This is a feature specific comment. This feature has an option that suggests var only in cases where the type is apparent from RHS of the assignment. Currently, it would suggest var for something like `Tuple.Create(1, true)`. Naturally, it should now support `(int, bool) s = (1, true)` cases. 

It could be just a matter of adding a test and making sure it works, or the feature may need some tweaking to get that case working. Eitherway, feel free to add a test and a workitem if further work is required.

jcouv replied:
I'll stop by to discuss.

jcouv replied:
I added a test for `ValueTuple<...> x = ValueTuple.Create(...);`. It works as expected (the type is apparent from the right-hand side).

But for `(int a, bool) x = (a: 1, true);` the type is not apparent. Filed issue #11094.

For `(int, int) x = ValueTuple.Create(1, 1);`, this also won't work yet. But I think it will work later this week due to some pending compiler changes. Filed issue #11095.

balajikris commented on src/EditorFeatures/Test/BraceHighlighting/AbstractBraceHighlightingTests.cs:
```diff
@@ -16,9 +17,9 @@ namespace Microsoft.CodeAnalysis.Editor.UnitTests.BraceHighlighting
     public abstract class AbstractBraceHighlightingTests
     {
-        protected async Task TestBraceHighlightingAsync(string markup)
+        protected async Task TestBraceHighlightingAsync(string markup, CSharpParseOptions options = null)
```
Hmm, can you please check if `AbstractBraceHighlightingTests` is the base class for C# and VB brace highlighting tests? If so, then, passing `CSharpParseOptions` to the base ctor doesn't seem right.

jcouv replied:
`AbstractBraceHighlightingTests` is only referenced in C# projects as it is. But I see your point, so I'll change to `ParseOptions` here.

balajikris commented on src/EditorFeatures/Test/BraceHighlighting/AbstractBraceHighlightingTests.cs:
```diff
@@ -42,6 +43,6 @@ public abstract class AbstractBraceHighlightingTests
         }
 
-        protected abstract Task<TestWorkspace> CreateWorkspaceAsync(string markup);
+        protected abstract Task<TestWorkspace> CreateWorkspaceAsync(string markup, CSharpParseOptions options);
```
same as above.

jcouv replied:
Done. Changed to `ParseOptions` to make room for future VB usage of this base class.

balajikris commented:
> Additional test suggestions within this range are appreciated.

I see, this is ongoing work.  you might also find following the [IDE test plan document](https://github.com/dotnet/roslyn/blob/master/docs/contributing/IDE%20Test%20Plan.md) to be useful to ensure you do not accidentally miss any features.

balajikris commented on src/EditorFeatures/CSharpTest/BraceMatching/CSharpBraceMatcherTests.cs:
```diff
@@ -747,5 +749,41 @@ static void Main(string[] args)
+            var code = @"public class C { $$(int, int, int, int, int, int, int, int) x; }";
+            var expected = @"public class C { (int, int, int, int, int, int, int, int[|)|] x; }";
+
+            await TestAsync(code, expected, TestOptions.Regular.WithTuplesFeature());
```
If you are testing this by `ParseOptions`, does `TestOptions.Regular.WithTuplesFeature()` test both regular C# and C# interactive cases? Just noting that if the ordinary flow tests both the regular and script options, then this feature specific flow should do the same.

jcouv replied:
Thanks. I will modify to cover scripting case as well.

jcouv commented:
@balajikris Thanks a lot for the review. I will go over the feedback and address tomorrow morning. 

I did follow the test plan manually. It was pretty useful to guide my exploration. 
I have filed issues based on what I found (here are all the [tuple issues](https://github.com/dotnet/roslyn/issues?q=is%3Aopen+label%3A%22New+Language+Feature+-+Tuples%22+sort%3Aupdated-desc)).

balajikris commented:
:+1:

jcouv commented:
@balajikris I pushed another commit which adds the coverage for script/interactive. Also adds a test we discussed for "Use Implicit Type" refactoring (I filed issues on that too).

@dotnet/roslyn-ide Please provide a second code review.

balajikris commented on src/EditorFeatures/CSharpTest/Diagnostics/UseImplicitOrExplicitType/UseImplicitTypeTests.cs:
```diff
@@ -1443,13 +1443,41 @@ static void M()
+
+            await TestAsync(before, after, options: ImplicitTypeEverywhere(), parseOptions: TestOptions.Regular.WithTuplesFeature(), withScriptOption: true);
+
+            await TestMissingAsync(before, options: ImplicitTypeWhereApparent(), parseOptions: TestOptions.Regular.WithTuplesFeature(), withScriptOption: true);
```
I think this test will fail, based on our offline conversation? you could check them in disabled and add a workitem attribute pointing to the bugs you filed.

balajikris replied:
Ah, TestMissingAsync, I see. In that case, adding the workitem would seem helpful because then we have context about why it started failing when someone fixes this.

jcouv replied:
No, this test passes. The second call is Test*Missing*Async.

balajikris commented:
Thanks for the tests and for filing bugs. :+1:  from me.

jcouv commented:
@VSadov I added the compiler test for TupleTypeSymbol defaulting to declaring itself as a struct when the underlying type is error.

VSadov commented:
LGTM

jcouv commented:
@dotnet-bot test prtest/lin/dbg/unit32 please

jcouv commented:
One more issue popped up when I pulled recent changes. I removed the test and added issue #11154 to track it, so that I don't block this PR.
In short, in `ValueTuple<int, int> s = ValueTuple.Create(1, 1);` the User Implicit Type refactoring no longer sees that the type is apparent on the right.
